lib/Sema/BuilderTransform.cpp

```diff
@@ -622,12 +622,104 @@ class BuilderClosureVisitor
         DeclNameLoc(endLoc), /*implicit=*/true);
   }
 
+  VarDecl *visitSwitchStmt(SwitchStmt *switchStmt) {
+    // Generate constraints for the subject expression, and capture its
+    // type for use in matching the various patterns.
+    Expr *subjectExpr = switchStmt->getSubjectExpr();
+    if (cs) {
+      // FIXME: Add contextual type purpose for switch subjects?
+      SolutionApplicationTarget target(subjectExpr, dc, CTP_Unused, Type(),
+                                       /*isDiscarded=*/false);
+      if (cs->generateConstraints(target, FreeTypeVariableBinding::Disallow)) {
+        hadError = true;
+        return nullptr;
+      }
```

Comment:
Do you need an explicit one-way constraint here so that we don't back-propagate type information from the cases into the subject expression?

Reply:
Great point! Right now, there isn't a one-way constraint separating the subject expression from the patterns it would match against, which is a difference from the way `switch` statements are normally type checked and will create more complicated constraint systems. I should absolutely capture the result of this expression in a one-way constraint to prevent that. Thank you!

Reply:
Fixed in the third commit, with a lovely example where we were accepting something we shouldn't.